transaction-pool/src/lib.rs

```diff
@@ -0,0 +1,153 @@
+use solana_sdk::pubkey::Pubkey;
+use std::collections::BTreeMap;
+use std::collections::HashMap;
+use std::collections::VecDeque;
+
+#[derive(Default, Clone, PartialEq, Debug)]
+pub struct Item {
+    pub priority: u64,
+    pub units: u64,
+    pub time: u64,
+    pub id: (u32, u32),
+}
+
+#[derive(Default, Clone)]
+pub struct Pool {
+    pub max_age: u64,
+    pub max_bucket_cu: u64,
+    pub max_block_cu: u64,
+    pub count: usize,
+    items: BTreeMap<u64, VecDeque<Item>>,
+}
+
+pub trait Table {
+    fn keys(&self, item: &Item) -> &[&Pubkey];
+}
+
+impl Pool {
+    pub fn insert(&mut self, item: Item) {
+        let bucket = self
+            .items
+            .entry(item.priority)
+            .or_insert_with(VecDeque::new);
+        bucket.push_back(item);
+        self.count += 1;
+    }
+    pub fn pop_block<T: Table>(&mut self, time: u64, table: &T) -> VecDeque<Item> {
+        let mut rv = VecDeque::new();
+        let mut total_cu: u64 = 0;
+        let mut block_full = false;
+        let mut buckets: HashMap<Pubkey, u64> = HashMap::new();
+        let mut gc = vec![];
+        for (k, v) in &mut self.items.iter_mut() {
```

Comment:
need revert iter in order to go from high priority to low?

Reply:
or we invert the value.

Reply:
rev is easy enough to keep it simple